src/error.rs

```diff
@@ -0,0 +1,385 @@
+use crate::NodePtr;
+use thiserror::Error;
+
+pub type Result<T> = std::result::Result<T, EvalErr>;
+
+#[derive(Debug, Error)]
+pub enum EvalErr {
+    #[error("Internal Error: {0}")]
+    InternalError(String),
+
+    #[error("Encoding / Decoding Error")]
+    SerializationError,
+
+    #[error("Out of Memory")]
+    OutOfMemory,
+
+    #[error("Cost Exceeded")]
+    CostExceeded,
+
+    #[error("Cost must be greater than zero.")]
+    CostBelowZero,
+
+    #[error("Too Many Pairs")]
+    TooManyPairs,
+
+    #[error("Too Many Atoms")]
+    TooManyAtoms,
+
+    #[error("Path Into Atom")]
+    PathIntoAtom,
+
+    #[error("Unknown Softfork Extension")]
+    UnknownSoftforkExtension,
+
+    #[error("Softfork specified cost mismatch")]
+    SoftforkSpecifiedCostMismatch,
+
+    #[error("clvm raise")]
+    Raise(NodePtr),
+
+    #[error("in ((X)...) syntax X must be lone atom")]
+    InPairMustBeLoneAtom(NodePtr),
```

Comment:
I think a lot of these error codes could be merged into a single `InvalidArg` error, with a `NodePtr` and a `String` describing the specific error to a human.

I think these could be turned into `InvalidArg`: `InPairMustBeLoneAtom`, `FirstOfNonCons`, `RestOfNonCons`, `ModPowNegativeExponent`, `RequiresIntArgument`, `RequiresPositiveIntArgument`, `RequiresInt32Args`, `RequiresInt64Args`, `RequiresArgs`, `TakesNoMoreThanArgs`, `RequiresAtom`, `UsedOnList`, `TakesExactlyArgs`, `ExpectedAtomGotPair`, `InvalidArgs2or3`, `InvalidIndices`, `ConcatOnList`, `NotValidG1Point`, `G1MapInvalidArgs`, `NotValidG2Point`, `G2MapInvalidArgs`, `NotG2Size`, `MessageDigestNot32Bytes`, `PubkeyNotValid`

Reply:
string not constant for each, basically add string arg to InvalidArg and go type by type.

Reply:
done.